### PR TITLE
CSRF exempt for WebhookView when authenticate via user/pass

### LINKED ISSUE #4986: csrf required for triggering build via http auth authenticated webhook
## Details

Hi,

In OpenStack, we've been using authentication with the webhook endpoints to trigger builds.

The way this has been working is that projects add a common ``openstackci`` user as an admin, and then in their project configuration inside our platform they add their webhook URL.

When new releases are made, our CI system uses the ``openstackci`` account (with the password only known to it) to authenticate and trigger RTD builds for the releasing project with a POST.  The exact code that does this is https://git.openstack.org/cgit/openstack-infra/zuul-jobs/tree/roles/trigger-readthedocs/tasks/main.yaml#n11

But in essence that boils down to (for, say jenkins-job-builder project)

```
curl --request POST -u openstackci:OURPASSWORD https://readthedocs.org/api/v2/webhook/jenkins-job-builder/47271/
```

Our jobs starting failing around 2018-12-05 (http://zuul.openstack.org/builds?job_name=trigger-readthedocs-webhook).  The jobs don't log the output of this POST so we don't leak the password into logs.

However, testing manually, the responses indicate a content-length is now required, and a referer, which ends up with

```
$ curl --request POST -H "Content-Length:0" -u openstackci:PASSWORD -e https://readthedocs.org/api/v2/webhook/jenkins-jo2/webhook/jenkins-job-builder/47271/ https://readthedocs.org/api/v2/webhook/jenkins-jo2/webhook/jenkins-job-builder/47271/
{"detail":"CSRF Failed: CSRF cookie not set."}
```

So I'm guessing all this points to CSRF has been somehow enabled for webhook endpoints when using HTTP basic-auth?  Is this possibly a new configuration?

We don't use the token-based webhook method because this system means we only have one common secret to keep -- the openstackci user password -- and everything else is self-service for projects who simply add the common user as an admin and include in their configuration their endpoint URL.

Any help with this will be greatly appreciated, thanks

### PR DESCRIPTION
**Pull request by humitos**
I think this bug was introduced at https://github.com/rtfd/readthedocs.org/commit/f95cae35bafbd8f7b80027199bf64e94e048dd70

Closes #4986

### CODE REVIEW

**Comment by agjohnson** [comment]
I'm waffling on this PR.

Removing this would likely(?) be opening ourselves up to CSRF attacks that use a logged in user's session to perform actions on our API. Without a CSRF token from RTD, a malicious site could use JS to interact with our API on behalf of the logged in user. We can mitigate this by requiring a CSRF token from RTD, in much the same way that we protect our dashboard forms against malicious JS. But i agree that this is extra burden on the user that makes for a worse UX.

This is one of the reasons why I lean towards divorcing user sessions/auth and API auth, though I suppose that offering token auth for a case like this removes the burden of CSRF tokens and session auth on API requests.

**Comment by humitos** [comment]
I agree with all you said here.

> Removing this would likely(?) be opening ourselves up to CSRF attacks that use a logged in user's session to perform actions on our API

I'm exempting _only_ the API view for the Webhooks not all the API endpoints. It's still "dangerous" because a malicious JS could trigger a crazy amount of builds, but it won't be able to _change anything_ in the project.

On the other hand, if we are going to not allow this, we need to update our docs at

https://docs.readthedocs.io/en/latest/webhooks.html#authentication

and reply to @ianw at #4986.

> This is one of the reasons why I lean towards divorcing user sessions/auth and API auth, though I suppose that offering token auth for a case like this removes the burden of CSRF tokens and session auth on API requests.

This is _the_ solution to me as well, and APIv3 will behave in this way.

**Comment by ianw** [comment]
> This is one of the reasons why I lean towards divorcing user sessions/auth and API auth, though I suppose that offering token auth for a case like this removes the burden of CSRF tokens and session auth on API requests.

By this you mean JWT type tokens for the user?  Rather than project-specific tokens?

As an infrastructure project hosting many projects that want to publish to RTD via us, it is much easier for us to run things via a single authenticated user and to publish.  It means for our users they simply add the CI user to their project (so it can publish) and add the publishing job with their non-secret project id to their CI release pipeline.  We only need to manage the one one secret that is the CI user password.  We could do a non-expiring token just the same.

We *can* take the alternative approach of each project using their per-project token to trigger builds, but it means managing many more secrets.  Each client project needs to then get their token, encrypt it and write a job to use it.  And because we don't want to leak the tokens we don't publish the logs, so then it becomes another step if debugging where projects probably need to get infrastructure admins involved.

**Comment by gorshunovr** [comment]
Team, please, review this patch. Thank you very much.

**Comment by humitos** [comment]
Upgrade the PR to solve conflicts. Ready to merge once tests pass.

File: readthedocs/restapi/views/integrations.py [hunk 1]
```diff
@@ -422,6 +422,11 @@ class WebhookView(APIView):
     be.
     """
 
+    # We want to avoid CSRF checking when authenticating by user/password on
+    # this API endpoint so we can make a request like:
+    # curl -X POST -d "branches=branch" -u user:pass -e URL /api/v2/webhook/test-builds/{pk}/
+    authentication_classes = [CsrfExemptSessionAuthentication]
```
Comment (ericholscher) [inline readthedocs/restapi/views/integrations.py]: I thought we were doing basic auth here, not session auth?

**Comment by stsewd** [comment]
I guess we can close this one. Problem fixed in https://github.com/rtfd/readthedocs.org/pull/5311

**Pull request closed without merging**